common/src/main/java/com/microsoft/identity/common/internal/cache/MsalCppOAuth2TokenCache.java

```diff
@@ -0,0 +1,151 @@
+
+import static com.microsoft.identity.common.exception.ErrorStrings.ACCOUNT_IS_SCHEMA_NONCOMPLIANT;
+import static com.microsoft.identity.common.exception.ErrorStrings.CREDENTIAL_IS_SCHEMA_NONCOMPLIANT;
+import static com.microsoft.identity.common.internal.cache.SharedPreferencesAccountCredentialCache.DEFAULT_ACCOUNT_CREDENTIAL_SHARED_PREFERENCES;
+import static com.microsoft.identity.common.internal.dto.CredentialType.ID_TOKEN_TYPES;
+
+/**
+ * Sub class of {@link MsalCppOAuth2TokenCache} to add specific public api's required for MSAL CPP library.
+ */
+public class MsalCppOAuth2TokenCache
+        <GenericOAuth2Strategy extends OAuth2Strategy,
+                GenericAuthorizationRequest extends AuthorizationRequest,
+                GenericTokenResponse extends TokenResponse,
+                GenericAccount extends BaseAccount,
+                GenericRefreshToken extends com.microsoft.identity.common.internal.providers.oauth2.RefreshToken>
+        extends MsalOAuth2TokenCache<
+        GenericOAuth2Strategy,
+        GenericAuthorizationRequest,
+        GenericTokenResponse,
+        GenericAccount,
+        GenericRefreshToken> {
+
+
+    /**
+     * Constructor of MsalOAuth2TokenCache.
+     *
+     * @param context                  Context
+     * @param accountCredentialCache   IAccountCredentialCache
+     * @param accountCredentialAdapter IAccountCredentialAdapter
+     */
+    private MsalCppOAuth2TokenCache(final Context context,
+                                   final IAccountCredentialCache accountCredentialCache,
+                                   final IAccountCredentialAdapter accountCredentialAdapter) {
+        super(context, accountCredentialCache, accountCredentialAdapter);
+    }
+
+    /**
+     * Factory method for creating an instance of MsalCppOAuth2TokenCache
+     * <p>
+     * NOTE: Currently this is configured for AAD v2 as the only IDP
+     *
+     * @param context The Application Context
+     * @return An instance of the MsalCppOAuth2TokenCache.
+     */
+    public static MsalCppOAuth2TokenCache create(@NonNull final Context context){
+        final MsalOAuth2TokenCache msalOAuth2TokenCache = MsalOAuth2TokenCache.create(context);
+        return new MsalCppOAuth2TokenCache(
+                context,
+                msalOAuth2TokenCache.getAccountCredentialCache(),
+                msalOAuth2TokenCache.getAccountCredentialAdapter()
+        );
+    }
+
+    /**
+     * @param accountRecord : AccountRecord associated with the input credentials.
+     * @param credentials   : list of Credential which can include AccessTokenRecord, IdTokenRecord and RefreshTokenRecord.
+     *                      Note : Both IdTokenRecord and RefreshTokenRecord need to be non null. AccessTokenRecord can be optional.
+     * @throws ClientException : If the supplied Account or Credential are null or schema invalid.
+     */
+    public void saveCredentials(@NonNull final AccountRecord accountRecord,
```

Comment:
@kreedula It occurs to me now that we ought to add some thread safety around this API -- if a few different threads start calling `removeRefreshTokenIfNeeded()` simultaneously we're going to have race conditions writing/reading from disk (saving new RTs while writing deleting old ones)